services/web/server/src/simcore_service_webserver/meta_core_repos.py

```diff
@@ -0,0 +1,155 @@
+        repo_id, offset=offset, limit=limit
+    )
+
+    checkpoints = [Checkpoint.from_commit_log(commit, tags) for commit, tags in logs]
+    assert len(checkpoints) <= limit if limit else True  # nosec
+    assert total_number_of_commits > 0
+
+    return checkpoints, total_number_of_commits
+
+
+async def create_checkpoint(
+    vc_repo: VersionControlRepository,
+    project_uuid: UUID,
+    *,
+    tag: str,
+    message: Optional[str] = None,
+) -> Checkpoint:
+    repo_id = await vc_repo.get_repo_id(project_uuid)
+    if repo_id is None:
+        repo_id = await vc_repo.init_repo(project_uuid)
+
+    commit_id = await vc_repo.commit(repo_id, tag=tag, message=message)
+    commit, tags = await vc_repo.get_commit_log(commit_id)
+    assert commit  # nosec
+
+    return Checkpoint.from_commit_log(commit, tags)
+
+
+async def get_checkpoint(
+    vc_repo: VersionControlRepository,
+    project_uuid: UUID,
+    ref_id: RefID,
+) -> Checkpoint:
+
+    repo_id, commit_id = await vc_repo.as_repo_and_commit_ids(project_uuid, ref_id)
+    assert repo_id  # nosec
+
+    commit, tags = await vc_repo.get_commit_log(commit_id)
+    return Checkpoint.from_commit_log(commit, tags)
+
+
+async def update_checkpoint(
+    vc_repo: VersionControlRepository,
+    project_uuid: UUID,
+    ref_id: RefID,
+    *,
+    message: Optional[str] = None,
+    tag: Optional[str] = None,
+) -> Checkpoint:
+
+    repo_id, commit_id = await vc_repo.as_repo_and_commit_ids(project_uuid, ref_id)
+
+    if message is None and tag is None:
+        log.warning(
+            "Nothing to update. Skipping updating ref %s of %s", ref_id, project_uuid
+        )
+    else:
+        await vc_repo.update_annotations(repo_id, commit_id, message, tag)
+
+    commit, tags = await vc_repo.get_commit_log(commit_id)
+    return Checkpoint.from_commit_log(commit, tags)
+
+
+async def checkout_checkpoint(
+    vc_repo: VersionControlRepository,
+    project_uuid: UUID,
+    ref_id: RefID,
+) -> Checkpoint:
+    repo_id, commit_id = await vc_repo.as_repo_and_commit_ids(project_uuid, ref_id)
+
+    # check if working copy has changes, if so, auto commit it
+    try:
+        commit_id = await vc_repo.checkout(repo_id, commit_id)
+    except CleanRequiredError:
+        log.info("Local changes found. Auto-commiting project %s", project_uuid)
+        await vc_repo.commit(repo_id, message="auto commit")
+        commit_id = await vc_repo.checkout(repo_id, commit_id)
+
+    commit, tags = await vc_repo.get_commit_log(commit_id)
+    return Checkpoint.from_commit_log(commit, tags)
+
+
+async def get_workbench(
+    vc_repo: VersionControlRepository,
+    project_uuid: UUID,
+    ref_id: RefID,
+) -> WorkbenchView:
+    repo_id, commit_id = await vc_repo.as_repo_and_commit_ids(project_uuid, ref_id)
+
+    content = await vc_repo.get_snapshot_content(repo_id, commit_id)
+    return WorkbenchView.from_orm(content)
+
+
+list_repos_safe = validate_arguments(list_repos, config=CFG)
```

Comment:
I guess this was not applied as a decorator because you needed both versions?

Reply:
yes